### PR TITLE
lnrpc: re-enable custom records

### PR DESCRIPTION
**Pull request by joostjager**
This PR re-enables custom record sending that was disabled in #3575.

It leaves the database serialization format as is, because there currently is no functional need to change it.

### CODE REVIEW

**Comment by joostjager** [comment]
@cfromknecht, we spoke offline about the risk that we wouldn't be able to deserialize even custom records from our own database as long as odd/even checking was part of the `tlv` package. That has been removed in #3701, but looking at the hop serialization code again it also seems that we weren't using tlv there in the first place?

**Comment by joostjager** [comment]
Added the custom range check. Hopefully this PR didn't miss any entry/exit point for custom records. Because of the duplicated send interface, it is quite a few places.

File: lnrpc/routerrpc/router_backend.go [hunk 1]
```diff
@@ -504,9 +536,11 @@ func (r *RouterBackend) extractIntentFromSendRequest(
...
 	}
 
-	var destTLV map[uint64][]byte
-	if len(destTLV) != 0 {
-		payIntent.FinalDestRecords = tlv.MapToRecords(destTLV)
+	payIntent.FinalDestRecords, err = UnmarshallCustomRecords(
+		rpcPayReq.DestTlv,
+	)
+	if err != nil {
+		return payIntent, err
```
Comment (guggero) [inline lnrpc/routerrpc/router_backend.go]: nit: returned value is ignored anyway, could also just be `nil, err`.

File: lnrpc/rpc.proto [hunk 2]
```diff
@@ -1854,6 +1854,14 @@ message QueryRoutesRequest {
...
     */
     uint32 cltv_limit = 11;
+
+    /** 
+    An optional field that can be used to pass an arbitrary set of TLV records
+    to a peer which understands the new records. This can be used to pass
+    application specific data during the payment attempt. If the destination
+    does not support the specified recrods, and error will be returned.
+    */
+    map<uint64, bytes> dest_tlv = 13;
```
Comment (guggero) [inline lnrpc/rpc.proto]: On every other field that has the type `bytes` involved, we have the following comment:
> When using REST, this field must be encoded as base64.

Do we want to add this here too?

File: lnrpc/routerrpc/router_backend.go [hunk 3]
```diff
@@ -364,90 +364,66 @@ func (r *RouterBackend) MarshallRoute(route *route.Route) (*lnrpc.Route, error)
...
+			return nil, fmt.Errorf("cannot decode pubkey %s",
+				rpcHop.PubKey)
+		}
+		copy(pubKeyBytes[:], pubKey)
+	} else {
 		// If no pub key is given of the hop, the local channel
 		// graph needs to be queried to complete the information
 		// necessary for routing.
-		return r.UnmarshallHopByChannelLookup(hop, prevNodePubKey)
+
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/router_backend.go]: merged

File: lnrpc/routerrpc/router_backend.go (hunk 1 above)
Comment (joostjager) [reply lnrpc/routerrpc/router_backend.go]: fixed

File: lnrpc/rpc.proto (hunk 2 above)
Comment (joostjager) [reply lnrpc/rpc.proto]: Done

File: lnrpc/routerrpc/router_backend.go [hunk 4]
```diff
@@ -358,99 +363,97 @@ func (r *RouterBackend) MarshallRoute(route *route.Route) (*lnrpc.Route, error)
...
 
-	var tlvRecords []tlv.Record
+	tlvRecords := tlv.MapToRecords(rpcRecords)
 
-	mpp, err := UnmarshalMPP(hop.MppRecord)
-	if err != nil {
-		return nil, err
+	// tlvRecords is sorted, so we only need to check that the first
+	// element is within the custom range.
+	if uint64(tlvRecords[0].Type()) < hop.CustomTypeStart {
```
Comment (cfromknecht) [inline lnrpc/routerrpc/router_backend.go]: do we know length > 0?

Comment (joostjager) [reply lnrpc/routerrpc/router_backend.go]: Maybe a zero element map from grpc is always nil, but not fully sure. Replaced the nil check at the top by a len()==0 check

File: lnrpc/rpc.proto (hunk 2 above)
Comment (cfromknecht) [reply lnrpc/rpc.proto]: To be consistent with the hop message, I would say this should be renamed to something like `dest_custom_records`. Comment should also mention custom record range

**Pull request merged**